algebra/src/tests/curves.rs

```diff
@@ -222,6 +237,319 @@ fn random_transformation_test<G: ProjectiveCurve>() {
+
+        assert_eq!(a, c);
+    }
+}
+
+pub fn random_batch_scalar_mul_test<G: ProjectiveCurve>() {
+    use algebra_core::curves::models::short_weierstrass_jacobian::{GroupAffine, GroupProjective};
+    use std::ops::MulAssign;
+    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+    for j in 0..ITERATIONS {
+        let size = std::cmp::min(1 << 7, 1 << (j + 4));
+        let mut a = Vec::with_capacity(size);
+        let mut s = Vec::with_capacity(size);
+
+        for i in 0..size {
+            a.push(G::rand(&mut rng));
+            s.push(G::ScalarField::rand(&mut rng));
+        }
+
+        let mut c = a.clone();
+        let mut t = s.clone();
+
+        let mut a: Vec<G::Affine> = a.iter().map(|p| p.into_affine()).collect();
+
+        let mut s: Vec<<G::ScalarField as PrimeField>::BigInt> =
+            s.iter().map(|p| p.into_repr()).collect();
+
+        a[..].batch_scalar_mul_in_place::<<G::ScalarField as PrimeField>::BigInt>(&mut s[..], 4);
+
+        for (p_c, s_t) in c.iter_mut().zip(t.iter()) {
+            p_c.mul_assign(*s_t);
+        }
+
+        let c: Vec<G::Affine> = c.iter().map(|p| p.into_affine()).collect();
+
+        for (p1, p2) in a.iter().zip(c) {
+            // println!("{}", *p1 == p2);
+            assert_eq!(*p1, p2);
+        }
+    }
+}
+
+fn batch_bucketed_add_test<C: AffineCurve>() {
+    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+    const MAX_LOGN: usize = 14;
+    let random_elems = create_pseudo_uniform_random_elems(&mut rng, MAX_LOGN);
+
+    for i in (MAX_LOGN - 4)..(ITERATIONS / 2 + MAX_LOGN - 4) {
+        let n_elems = 1 << i;
+        let n_buckets = 1 << (i - 3);
+
+        let mut bucket_assign = Vec::<usize>::with_capacity(n_elems);
+        let step = Uniform::new(0, n_buckets);
+
+        for _ in 0..n_elems {
+            bucket_assign.push(step.sample(&mut rng));
+        }
+
+        let mut res1 = vec![];
+        let mut elems_mut = random_elems[0..n_elems].to_vec();
+        let now = std::time::Instant::now();
+        res1 = batch_bucketed_add::<C>(n_buckets, &mut elems_mut[..], &bucket_assign[..]);
+        println!(
+            "batch bucketed add for {} elems: {:?}",
+            n_elems,
+            now.elapsed().as_micros()
+        );
+
+        let mut res2 = vec![C::Projective::zero(); n_buckets];
+        let mut elems = random_elems[0..n_elems].to_vec();
+
+        let now = std::time::Instant::now();
+        for (&bucket_idx, elem) in bucket_assign.iter().zip(elems) {
+            res2[bucket_idx].add_assign_mixed(&elem);
+        }
+        println!(
+            "bucketed add for {} elems: {:?}",
+            n_elems,
+            now.elapsed().as_micros()
+        );
+
+        let res1: Vec<C::Projective> = res1.iter().map(|&p| p.into()).collect();
+
+        for (i, (p1, p2)) in res1.iter().zip(res2).enumerate() {
+            assert_eq!(*p1, p2);
+        }
+    }
+}
+
+macro_rules! batch_verify_test {
```

Comment:
Are the commented out tests take too long to run?
Btw, the printing ones can remain, we usually don't see the prints.

Reply:
The test below is just non deterministic and only works for small security parameters. It's an anti-test. I'd like to leave it there in case anyone wants to do sanity checking.

The recursive stuff I will put in an issue and we can add it back.

Reply:
Got it. Maybe we can put a comment about that?

Reply:
I've just seen you did, thanks!